File: pyshell/utils/heredocs.py
#!/usr/bin/env python3

import utils.global_var as gv
import utils.key as k

# import utils.tagstokens as tt
# Should be this next import statement but create circular import error.
# from utils.tagstokens import TagsTokens


def replace_first_tab(tok):
    i = 0
    len_tok = len(tok)
    while i < len_tok and ord(tok[i]) == 9:
        i += 1
    return tok[:i]


class Heredocs():
    def __init__(self, key, len_key, minus):
        import utils.tagstokens as tt
        self.key = key
        self.quoted = k.quoted_gold_key(key)
        self.len_key = len_key
        self.minus = minus
        self.closed = False
        self.tagstokens = tt.TagsTokens()
        self.stack = []
        self.in_cmdsubst = False

    def add_tags_tokens(self, tag, token):
        if self.quoted:
            tag = 'STMT'
        elif not self.in_cmdsubst:
            if tag in gv.GRAMMAR.dquotes_opening_tags:
                self.in_cmdsubst = True
                self.stack.append(gv.GRAMMAR.dquotes_opening_tags[tag])
            else:
                tag = tag if tag in ['VAR', 'STMT',
                                     'SPACES', 'NEW_LINE'] else 'STMT'
        else:
            if tag == self.stack[-1]:
                self.stack.pop(-1)
                if self.stack == []:
                    self.in_cmdsubst = False
            elif tag == 'QUOTE' and self.stack[-1] != 'DQUOTES':
                self.stack.append(tag)
            elif tag == 'DQUOTES' and self.stack[-1] != 'QUOTE':
                self.stack.append(tag)
            elif tag in gv.GRAMMAR.opening_tags and self.stack[-1] != 'QUOTE':
                if self.stack[-1] == 'DQUOTES' and \
                        tag in gv.GRAMMAR.dquotes_opening_tags:
                    self.stack.append(gv.GRAMMAR.dquotes_opening_tags[tag])
                else:
                    self.stack.append(gv.GRAMMAR.opening_tags[tag])
        self.tagstokens.append(tag, token)

    def striplines(self):
        i = 0
        new_line = True
        tag = ''
        tok = ''
        while i < self.tagstokens.length:
            tok, tag = self.tagstokens[i]
            if tag == 'NEW_LINE':
                new_line = True
            elif tag == 'SPACES' and new_line:
                self.tagstokens.tokens[i] = replace_first_tab(tok)
            else:
                new_line = False
            i += 1

    def close(self):
        self.tagstokens = self.tagstokens.copytt(0, -self.len_key)
        if self.minus:
            self.striplines()
        self.tagstokens.check_syntax(True)
        self.closed = True
        # self.tagstokens.incomplete = False

    def __str__(self):
        str0 = '\nHEREDOC: {} | closed: {} | Minus: {}\n'.format(
            self.key, self.closed, self.minus)
        str0 += 'DOC:\n{}\n'.format(str(self.tagstokens))
        return str0

File: src/ast_build/heredocs.c
#include "tmpsh.h"
#include "libft.h"
#include "tagstokens.h"
#include "check_syntax_tagstokens.h"

int			free_heredocs(t_heredocs **self, int status)
{
	ft_strdel(&(*self)->key);
	free_tagstokens(&(*self)->tagstokens, status);
	ft_memdel((void **)self);
	return (status);
}

void	init_heredocs(t_heredocs **self, char *key, int len_key, t_bool minus)
{
	*self = ft_memalloc(sizeof(t_heredocs));
	(*self)->key = ft_strdup(key);
	(*self)->quoted = quoted_gold_key(key);
	(*self)->len_key = len_key;
	(*self)->minus = minus;
	(*self)->closed = false;
	init_tagstokens(&(*self)->tagstokens, NULL, NULL);
	(*self)->in_cmdsubst = false;
}

void	check_stack(t_heredocs *self, char *tag)
{
	char		*last_stack;
	char		*tmp_tag;

	last_stack = vindex_pylst(self->stack, -1);
	if (ft_strequ(tag, last_stack))
	{
		pop_pylst(&self->stack, -1);
		if (len_pylst(self->stack) == 0)
			self->in_cmdsubst = false;
	}
	else if (ft_strequ(tag, "QUOTE") && !ft_strequ(last_stack, "DQUOTES"))
		push_pylst(&self->stack, tag, 0, _ptr);
	else if (ft_strequ(tag, "DQUOTES") && !ft_strequ(last_stack, "QUOTE"))
		push_pylst(&self->stack, tag, 0, _ptr);
	else if (search_value(g_grammar->opening_tags, tag) && \
			!ft_strequ(last_stack, "QUOTE"))
	{
		if (ft_strequ(last_stack, "DQUOTES") && \
				(tmp_tag = search_value(g_grammar->dquotes_opening_tags, tag)))
			push_pylst(&self->stack, tmp_tag, 0, _ptr);
		else
		{
			tmp_tag = search_value(g_grammar->opening_tags, tag);
			push_pylst(&self->stack, tmp_tag, 0, _ptr);
		}
	}
}

void	add_tags_tokens(t_heredocs *self, char *tag, char *token)
{
	char	*tmp_tag;

	if (self->quoted)
		tag = "STMT";
	else if (!self->in_cmdsubst)
	{
		if ((tmp_tag = search_value(g_grammar->dquotes_opening_tags, tag)))
		{
			self->in_cmdsubst = true;
			push_pylst(&self->stack, tmp_tag, 0, _ptr);
		}
		else
			tag = in(tag, "VAR", "STMT", "SPACES", "NEW_LINE", NULL) ?\
				  tag : "STMT";
	}
	else
		check_stack(self, tag);
	append_tagstokens(tag, token);
}

void	striplines(t_heredocs *self)
{
	size_t		index;
	t_bool	new_line;
	char	*tag;
	char	*token;

	index = 0;
	new_line = true;
	while (index < self->tagstokens->length)
	{
		token = vindex_pylst(self->tagstokens->tokens, index);
		tag = vindex_pylst(self->tagstokens->tags, index);
		if (ft_strequ(tag, "NEW_LINE"))
			new_line = true;
		else if (ft_strequ(tag, "SPACES") && new_line)
		{
			//replace_first_tab seem to not exist.
			//Check allocation, set _ptr by default.
			//update_pylst(self->tagstokens->tokens, index, \
			//		replace_first_tab(token), 0, _ptr);
		}
		else
			new_line = false;
		index++;
	}
}

void	close_heredoc(t_heredocs *self)
{
	//self->tagstokens = copytt(self, 0, -self->len_key);//Seem to not exist
	self->tagstokens->not_heredocs = false;
	if (self->minus)
		striplines(self);
	check_syntax_tagstokens(self->tagstokens); //Check if check_stack in python is check_syntax_tagstokens
	self->closed = true;
}

static char	*quick_join(char *s1, char *s2, t_bool free_s1, t_bool free_s2)
{
	return (ft_fstrjoin(&s1, &s2, free_s1, free_s2));
}

char	*str_heredocs(t_heredocs *self)
{
	char	*str;
	char	*bool_str;

	str = quick_join("\nHEREDOC", self->key, false, false);
	str = quick_join(str, " | closed: ", true, false);
	bool_str = self->closed ? "True" : "False";
	str = quick_join(str, bool_str, true, false);
	str = quick_join(str, " | Minus: ", true, false);
	bool_str = self->minus ? "True" : "False";
	str = quick_join(str, bool_str, true, false);
	str = quick_join(str, "\n", true, false);
	str = quick_join(str, "DOC:\n", true, false);
	str = quick_join(str, str_tagstokens(self->tagstokens), true, true);
	return (str);
}
